### PR TITLE
Move rails based pipeline config page to typescript spa

### PR DESCRIPTION
**Pull request by bdpiparva**
Move rails based pipeline config page to typescript spa

### CODE REVIEW

File: api/api-shared-v10/src/main/java/com/thoughtworks/go/apiv10/admin/shared/representers/stages/ApprovalRepresenter.java [hunk 1]
```diff
@@ -33,9 +33,7 @@ public static void toJSON(OutputWriter jsonWriter, Approval approval) {
         }
 
         jsonWriter.add("type", approval.getType());
-        if (approval.isAllowOnlyOnSuccess()) {
-            jsonWriter.add(Approval.ALLOW_ONLY_ON_SUCCESS, approval.isAllowOnlyOnSuccess());
-        }
+        jsonWriter.add(Approval.ALLOW_ONLY_ON_SUCCESS, approval.isAllowOnlyOnSuccess());
```
Comment (GaneshSPatil) [inline api/api-shared-v10/src/main/java/com/thoughtworks/go/apiv10/admin/shared/representers/stages/ApprovalRepresenter.java]: why is this change required?

File: api/api-shared-v9/src/main/java/com/thoughtworks/go/apiv9/admin/shared/representers/stages/ApprovalRepresenter.java [hunk 2]
```diff
@@ -33,9 +33,7 @@ public static void toJSON(OutputWriter jsonWriter, Approval approval) {
         }
 
         jsonWriter.add("type", approval.getType());
-        if (approval.isAllowOnlyOnSuccess()) {
```
Comment (GaneshSPatil) [inline api/api-shared-v9/src/main/java/com/thoughtworks/go/apiv9/admin/shared/representers/stages/ApprovalRepresenter.java]: why is this change required?

File: server/src/main/webapp/WEB-INF/rails/webpack/jsx/mithril-component.ts [hunk 3]
```diff
@@ -37,6 +37,7 @@ export abstract class MithrilViewComponent<Attrs = {}> implements m.ClassCompone
 
 export interface RestyleAttrs<T> {
   [key: string]: any;
+
```
Comment (GaneshSPatil) [inline server/src/main/webapp/WEB-INF/rails/webpack/jsx/mithril-component.ts]: whitespace?

File: api/api-shared-v10/src/main/java/com/thoughtworks/go/apiv10/admin/shared/representers/stages/ApprovalRepresenter.java (hunk 1 above)
Comment (bdpiparva) [reply api/api-shared-v10/src/main/java/com/thoughtworks/go/apiv10/admin/shared/representers/stages/ApprovalRepresenter.java]: IMO, we should always serialize the boolean values. This will keep the payload consistent. I don't see point why we should not remove the property when value is set to false.

File: server/src/main/webapp/WEB-INF/rails/webpack/jsx/mithril-component.ts (hunk 3 above)
Comment (bdpiparva) [reply server/src/main/webapp/WEB-INF/rails/webpack/jsx/mithril-component.ts]: It is auto indentation by idea.

File: server/src/main/webapp/WEB-INF/rails/webpack/models/pipeline_configs/artifact.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+export abstract class Artifact {
+  readonly type = Stream<ArtifactType>();
+
+  protected constructor(type: ArtifactType) {
+    this.type(type);
+  }
+
+  static fromJSON(json: ArtifactJSON) {
+    switch (json.type) {
+      case "test":
```
Comment (GaneshSPatil) [inline server/src/main/webapp/WEB-INF/rails/webpack/models/pipeline_configs/artifact.ts]: how about `json.type` an enum?

Comment (bdpiparva) [reply server/src/main/webapp/WEB-INF/rails/webpack/models/pipeline_configs/artifact.ts]: Yes, Can convert it to enum.

**Comment by bdpiparva** [comment]
Closing in favour of https://github.com/gocd/gocd/pull/7727
